### PR TITLE
retry on indexer failure

### PR DESCRIPTION
**Pull request by apoorvsadana**

### CODE REVIEW

**Review by EvolveArt — COMMENTED** [review]
Wondering if a better design would not be to have a static Hashmap that keep tracks of the attempts for each indexer id, wdyt ?

File: src/handlers/indexers/indexer_types/mod.rs [hunk 1]
```diff
@@ -77,8 +86,24 @@ pub trait Indexer {
...
-                                publish_failed_indexer(Uuid::parse_str(indexer_id.as_str()).expect("Invalid UUID for indexer")).await.unwrap();
+                                let indexer_id = Uuid::parse_str(indexer_id.as_str()).expect("Invalid UUID for indexer");
+                                let indexer_end_time = Utc::now().time();
+                                let indexer_duration = indexer_end_time - indexer_start_time;
+                                if indexer_duration.num_minutes() > 5 {
+                                    // if the indexer ran for more than 5 minutes, we will try to restart it
+                                    // with attempt id 1. we don't want to increment the attempt id as this was
+                                    // a successful run and a we want MAX_INDEXER_START_RETRIES to restart the indexer
+                                    tracing::error!("Indexer {} ran for more than 30 seconds, trying restart", indexer_id);
+                                    publish_start_indexer(indexer_id, 1).await.unwrap();
```
[the remark this answers is not in the document]
Comment (apoorvsadana) [reply src/handlers/indexers/indexer_types/mod.rs:97]: I think it's safe to unwrap here as it happens inside a separate thread so we can't propagate it either

File: src/handlers/indexers/indexer_types/mod.rs [hunk 2]
```diff
@@ -77,8 +86,24 @@ pub trait Indexer {
                             },
                             false => {
                                 tracing::error!("Child process exited with an error {}", indexer_id);
-                                // TODO: safe to unwrap here?
-                                publish_failed_indexer(Uuid::parse_str(indexer_id.as_str()).expect("Invalid UUID for indexer")).await.unwrap();
+                                let indexer_id = Uuid::parse_str(indexer_id.as_str()).expect("Invalid UUID for indexer");
```
[the remark this answers is not in the document]
Comment (apoorvsadana) [reply src/handlers/indexers/indexer_types/mod.rs:89]: Same as above, it's inside a separate thread so IndexerError won't help here

**Comment by apoorvsadana** [comment]
> Wondering if a better design would not be to have a static Hashmap that keep tracks of the attempts for each indexer id, wdyt ?

It's usually better to track attempts like this on the DB or queues because that is shared across all instances. If in the future, say we parallize the service to run on multiple instances, the HashMap will be local to each instance. So in the worst case the number of attempts would be proportional to the number of instances running.

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**